### ISSUE TITLE
CSS property extraction: deal with "double" font-display dfn

### ISSUE THREAD

**Issue by tidoust** [w3c/reffy#642]
The CSS extracts that Reffy produces assume that descriptor names are global identifiers. That is true for virtually all descriptors, but working on #641, I realized that the `font-display` descriptor is defined twice:
1. [`font-display` for the `@font-face` rule](https://drafts.csswg.org/css-fonts-4/#descdef-font-face-font-display)
2. [`font-display` for the `@font-feature-values` rule](https://drafts.csswg.org/css-fonts-4/#descdef-font-feature-values-font-display)

The definitions are essentially the same, except for the rule.

Since Reffy did not properly handle duplication of dfns, the second definition overrides the first definition in the extract, leading to:
https://github.com/w3c/webref/blob/master/ed/css/css-fonts.json#L439

Question is: what should the extract contain?

We cannot duplicate the keys since they are keys... One option is to merge the `for` property and create an array instead, as in:

```json
"font-display": {
  "name": "font-display",
  "for": ["@font-face", "@font-feature-values"],
  "value": "auto | block | swap | fallback | optional",
  "initial": "auto"
}
```

This looks good, but I wonder whether we should anticipate that there may be descriptors with the same name and different semantics for different rules in the future. Another option is to keep the definitions side-by-side:

```json
"font-display": [
  {
    "name": "font-display",
    "for": "@font-face",
    "value": "auto | block | swap | fallback | optional",
    "initial": "auto"
  },
  {
    "name": "font-display",
    "for": "@font-feature-values",
    "value": "auto | block | swap | fallback | optional",
    "initial": "auto"
  }
]
```

A third option is to stop using descriptor names as keys, and rather return an array of descriptors instead of an object indexed by descriptor names.

All of these options break the format of the extracts. A *less breaking* (but still somewhat breaking) alternative would be to use a comma-separated value list for the `for` property as in:

```json
"font-display": {
  "name": "font-display",
  "for": "@font-face, @font-feature-values",
  "value": "auto | block | swap | fallback | optional",
  "initial": "auto"
}
```

I personally find that last approach less satisfactory.

@dontcallmedom, thoughts?

**Comment by dontcallmedom** [comment]
I agree that the comma-separated approach feels brittle, since it builds on the assumption that a given descriptor name will always map to the same set of values - unless we can back that by a commitment in one of the CSS meta specs? Plus, as you say, it's still a breaking change in terms of format expectations.

I don't have strong feelings between the array of name-indexed dictionaries vs a name-indexed dictionary of arrays. I would guess that the latter is a bit less of a foot gun from the consumer side, since I could imagine code  using a `descriptors.find(x => x.name = 'font-display')` rather than `descriptors.filter(x => x.name = 'font-display')` (assuming mistakenly but most-of-the-time-rightly that there is only one such entry), whereas if you always get an array back from `descriptors[x]`, it's harder to ignore the expectation that you need to deal with multiple items.

**Comment by tabatkins** [comment]
There is definitely no such commitment, and in fact an anti-commitment: descriptors are definitely local to a given at-rule, and if this is the only collision so far that's purely by accident. These need to be tracked as completely separate definitions.

(There *should* be more conflicts that that, too; for example @font-face/src and @color-profile/src, which definitely have distinct grammars.)

**Comment by tidoust** [comment]
> There is definitely no such commitment, and in fact an anti-commitment: descriptors are definitely local to a given at-rule, and if this is the only collision so far that's purely by accident. These need to be tracked as completely separate definitions.

That makes sense. Thanks for the confirmation.

> (There *should* be more conflicts that that, too; for example @font-face/src and @color-profile/src, which definitely have distinct grammars.)

The bug only affect cases where the descriptor is defined more than once in the *same* spec and `font-display` seems to be the only occurrence so far. `@font-face/src` and `@color-profile/src` are defined in different specs, and thus appear in different extract files.

**Comment by OnkarRuikar** [comment]
There is one more case `font-family`.
[`@font-palette-values/font-family`](https://drafts.csswg.org/css-fonts/#font-family-prop):
```
{
  "name": "@font-palette-values/font-family",
  "syntax": "<family-name>#",
 }
```
[`@font-face/font-family`](https://drafts.csswg.org/css-fonts/#font-family-desc):
```
{
  "name": "@font-face/font-family",
  "syntax": "<family-name>",
 }
```

**Comment by tidoust** [comment]
Not sure if you're looking for a full list or reporting a bug in the extracted data, @OnkarRuikar.

If you're looking for a full list, you can look at [CSS extracts in Webref](https://github.com/w3c/webref/tree/main/ed/css), and extract descriptors under at-rules definitions to get those that are for different at-rules. See below for Node.js code that would do that. Today, that list seems to be:

- `aspect-ratio` defined for `@container` (in css-contain-3), `@media` (in mediaqueries-5)
- `font-display` defined for `@font-face` (in css-fonts), `@font-feature-values` (in css-fonts)
- `font-family` defined for `@font-face` (in css-fonts), `@font-palette-values` (in css-fonts)
- `height` defined for `@container` (in css-contain-3), `@media` (in mediaqueries-5)
- `orientation` defined for `@container` (in css-contain-3), `@media` (in mediaqueries-5)
- `src` defined for `@color-profile` (in css-color-5), `@font-face` (in css-fonts)
- `width` defined for `@container` (in css-contain-3), `@media` (in mediaqueries-5)

If you find another instance, that means there's either a bug in the extraction or in the spec!

Sample code to get to that result:

```js
const css = require('@webref/css');
css
  .listAll()
  .then(files => {
    let descriptors = {};
    for (const [shortname, data] of Object.entries(files)) {
      for (const atrule of data.atrules) {
        for (const { name } of atrule.descriptors) {
          if (!descriptors[name]) {
            descriptors[name] = [];
          }
          // Some at-rules and descriptors are re-defined in subsequent levels
          // (as any-hover in mediaqueries and mediaqueries-5), get rid of these
          if (!descriptors[name].find(rule => rule.name === atrule.name)) {
            descriptors[name].push({
              name: atrule.name,
              shortname: shortname
            });
          }
        }
      }
    }
    // Report descriptors defined for more than one at-rule
    for (const name of Object.keys(descriptors).sort()) {
      const atrules = descriptors[name].map(atrule => `\`${atrule.name}\` (in ${atrule.shortname})`);
      if (atrules.length > 1) {
        console.log(`- \`${name}\` defined for ${atrules.join(', ')}`);
      }
    }
  });
```
